### PR TITLE
Handle global classes when resolving property location

### PR DESCRIPTION
**Pull request by HolonProduction**
Fixes partially #74888
Fixes partially #71296 
Fixes partially #78454

Fixes partially #77643

Fixes partially godotengine/godot-proposals#8545 (Should be closed if #85215 is merged)

### CODE REVIEW

**Comment by YuriSizov** [comment]
> (partially)

If all of these issues are not fully fixed, then please put this between `Fixes` and the issue number, so that we don't auto-close them.

File: modules/gdscript/gdscript_editor.cpp [hunk 1]
```diff
@@ -552,6 +552,19 @@ static int _get_property_location(StringName p_class, StringName p_property) {
...
+	int depth = 0;
+	Ref<Script> scr = p_script;
+	while (scr.is_valid()) {
+		if (scr->get_member_line(p_property) != -1) {
+			return depth | ScriptLanguage::LOCATION_PARENT_MASK;
+		}
+		depth++;
+		scr = scr->get_base_script();
+	}
+	return depth + _get_property_location(p_script->get_class_name(), p_property);
```
[the remark this answers is not in the document]
Comment (HolonProduction) [reply modules/gdscript/gdscript_editor.cpp]: Good catch. I'd bet that's not the only place where those got messed up.

**Review by vnen — APPROVED** [review]
LGTM

File: modules/gdscript/gdscript_editor.cpp [hunk 2]
```diff
@@ -567,6 +580,19 @@ static int _get_constant_location(StringName p_class, StringName p_constant) {
 	return depth | ScriptLanguage::LOCATION_PARENT_MASK;
 }
 
+static int _get_constant_location(Ref<Script> p_script, StringName p_constant) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:583]:
```suggestion
static int _get_constant_location(Ref<Script> p_script, const StringName &p_constant) {
```

File: modules/gdscript/gdscript_editor.cpp [hunk 3]
```diff
@@ -582,6 +608,19 @@ static int _get_signal_location(StringName p_class, StringName p_signal) {
 	return depth | ScriptLanguage::LOCATION_PARENT_MASK;
 }
 
+static int _get_signal_location(Ref<Script> p_script, StringName p_signal) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:611]:
```suggestion
static int _get_signal_location(Ref<Script> p_script, const StringName &p_signal) {
```

**Comment by YuriSizov** [comment]
Needs a rebase though!

File: modules/gdscript/gdscript_editor.cpp [hunk 4]
```diff
@@ -597,6 +636,19 @@ static int _get_method_location(StringName p_class, StringName p_method) {
 	return depth | ScriptLanguage::LOCATION_PARENT_MASK;
 }
 
+static int _get_method_location(Ref<Script> p_script, StringName p_method) {
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:639]:
```suggestion
static int _get_method_location(Ref<Script> p_script, const StringName &p_method) {
```

**Review by AThousandShips — APPROVED** [review]

File: modules/gdscript/gdscript_editor.cpp [hunk 5]
```diff
@@ -552,6 +552,19 @@ static int _get_property_location(StringName p_class, StringName p_property) {
 	return depth | ScriptLanguage::LOCATION_PARENT_MASK;
 }
 
+static int _get_property_location(Ref<Script> p_script, StringName p_property) {
```
[the remark this answers is not in the document]
Comment (HolonProduction) [reply modules/gdscript/gdscript_editor.cpp:555]: Should we also change the signature for the other overloads?

[the remark this answers is not in the document]
Comment (AThousandShips) [reply modules/gdscript/gdscript_editor.cpp:555]: I'm not sure either way, it's not critical and would add unrelated changes, but might be worth it

[the remark this answers is not in the document]
Comment (HolonProduction) [reply modules/gdscript/gdscript_editor.cpp:555]: Apparently it was already done for the other overloads in #51156. A rebase will do.

**Comment by HolonProduction** [comment]
> Needs a rebase though!

Done

**Comment by YuriSizov** [comment]
@HolonProduction Is this the last PR needed to fully address the linked issues?

**Comment by HolonProduction** [comment]
> @HolonProduction Is this the last PR needed to fully address the linked issues?

https://github.com/godotengine/godot-proposals/issues/8545 can be closed (preferably as "not planed")

https://github.com/godotengine/godot/issues/74888, https://github.com/godotengine/godot/issues/71296 and https://github.com/godotengine/godot/issues/78454 are already closed because another fix made the code path from this PR irrelevant to them.

https://github.com/godotengine/godot/issues/77643 consists of multiple points, some of which still need to be addressed.

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**